### ISSUE TITLE
Develop spec of raw (compressed, encoded) transactions messages

### ISSUE THREAD

**Issue by hrharder** [ParadigmFoundation/ParadigmCore#48]
# Overview

The format of `order`, `rebalance`, and `witness` transaction types (one format used for all raw messages) is currently underspecified in documentation and spec. The current decoding process 

## Raw transaction format

All validator-signed transactions have the following outer-level schema (after decoding/decompression):

```js
// JSON shown (in JS)

const decodedSignedTx = {
  "type": "" // witness, order, or rebalance
  "data": {} // tx-type specific schema
  "proof": {
    "valPubKey": "" // reported validator public key
    "signature": "" // ed25519 signature (hex str)
  }
}
```

Where `proof.signature` is an ed25519 signature of the raw JSON-encoded `data` bytes (another to-do: change to signature of a hash) .

Functions found in `utils` and `core/utils` allow decoding and parsing of raw messages along with other utilities. 

### Encoding process

Transactions originate by calls to [`core/utils/TxGenerator`](https://github.com/paradigmfoundation/paradigmcore/blob/dev/src/core/util/TxGenerator.ts) instances, where raw transaction messages (see above) are encoded as follows (summarized).

1. Construct `tx.data` object depending on `tx.type`
1. Generate `message` from "raw" transaction bytes (TODO: deterministic serialization)
   - Generate JSON string of transaction object
   - `Buffer` string with base64 encoding
1. Sign `message` byte array using the `ed25519` signature specification and library.
   - Store `hex` encoded byte array in `tx.proof.signature` resultant signature 
   - Append `base64` encoded validator public key as `tx.proof.valPubKey`
1. Serialize (see above) entire transaction object (for `bigint` related reasons)
   - Generate a `Buffer` from the `utf8` serialized string
   - Compress the byte array using `zlib`
   - Serialize the compressed byte array as a `base64`-encoded `utf8` string
1. The final serialized, compressed, signed `base64` encoded string is delivered to the `core` state machine by ABCI server.

## Decoding process (in ABCI application)

The current decoding process implementation can be found [here.](https://github.com/paradigmfoundation/paradigmcore/blob/dev/src/crypto/PayloadCipher.ts#L126)

In summary:

1. Receive raw `RequestDeliverTx` bytes from ABCI server 
1. Convert the raw bytes to a utf8 encoded string (a base64 representation)
1. Decode the string as bytes64
1. Decompress compress raw tx using `zlib`

# Action items

- [ ] Develop robust spec of entire transaction/ABCI message generation/encoding pipeline
- [ ] Adjust ParadigmCore PoC implementation where logical

**Comment by hrharder** [comment]
Potentially a better JS implementation that allows for arbitrary has function: https://github.com/miguelmota/merkletreejs

**Comment by gchaincl** [comment]
Serialization:
Protobuf does not ensure deterministic serialization across languages [1].

Deterministic serialization across languages is vital if we want to have 2+ implementations. Otherwise, Tendermint's replay protection won't work (replay protection rely on transactions hash uniqueness).[5]

Tendermint uses a subset of Protobuf, `go-amino`[3], one of their goals is to provide deterministic serialization. Unfortunately, there's no port to other languages.

So, should we invest time in working on 2+ implementations?
Pros:
  - Having more implementations running the consensus protocol minimize failures rate when one of them is buggy; _run multiple implementations of the consensus code, and only accept a transaction as finalized if all of the implementations accept it_[4].
  - Assuming we design test specs, running them against both implementations will give us more certainty about the correctness of the implementation.
Cons:
  - It's not trivial (by trivial I mean, we can't just use a library, we'll need to implement it)

Considering that the main source of non-deterministic serialization in Protocol Buffers is maps [2], [6], we could consider using it avoiding maps; this solution is also suggested by MsgPack's spec [7]

Another motivation for finding a new way to serialize is efficiency, in this regard Protobuf is efficient 

Considering this I think we should:
- Evaluate how does Protobuf perform avoiding Maps.
- Consider MsgPack [7]
- Consider writing a binary format (as Bitcoin, Ethereum, and other does)

1: https://github.com/protocolbuffers/protobuf/releases/tag/v3.0.0
2: https://gist.github.com/kchristidis/39c8b310fd9da43d515c4394c3cd9510
3: https://github.com/tendermint/go-amino
4: https://blog.ethereum.org/2016/05/09/on-settlement-finality/
5: https://tendermint.com/docs/app-dev/app-development.html#replay-protection
6: https://havoc.io/post/deterministic-protobuf/
7: https://github.com/msgpack/msgpack/blob/master/spec.md#profile

**Comment by gchaincl** [comment]
Merkle Trees:

Tendermint block header includes a number of hashes `AppHash` is App specific. This Hash is set on the `Commit`[1] ABCI operation.

Merkle trees are useful to implement lite clients, they provide proofs of inclusion (and absence) to clients that don't hold the full chain but only the hashes.

It's up to the ABCI to build the tree and return the root hash on every `Commit` and up to the lite client to perform the verification. The proofs are included in the Query response when requested (see`prove` argument [2]).

As `AppHash` is not computed by tendermint we can use any implementation of merkle trees,

1: https://tendermint.com/docs/spec/abci/abci.html#commit
2: https://tendermint.com/docs/spec/abci/abci.html#query

**Comment by gchaincl** [comment]
Here's my attempt to use IAVL implementation inside our ABCI app:
https://gist.github.com/gchaincl/d051aa8fade4bb433782d43fc5282a05

For demo purposes, a key (`/key-<N>`) is being added to the Merkle Tree on every `Commit`, but they should be added on `DeliverTx` instead.

To test this, run the code `go run main.go`
And perform a query:
```
curl 'localhost:26657/abci_query?data="/key-1"'
{
  "jsonrpc": "2.0",
  "id": "",
  "result": {
    "response": {
      "key": "L2tleS0x",
      "value": "L3ZhbC0x",
      "proof": {
        "ops": [
          {
            "type": "iavl:v",
            "key": "L2tleS0x",
            "data": "rQIKqgIKKAgOED0YPSog7hb4Hr0MYJvT2rbZLxL4oOaZ9GbJPm6zf6+dBZfaNC8KKAgKEBQYIiog3+BM/jWAZY7LEW6ISeX/SqzJLLTI+4iAeRR+HkjHeY8KKAgIEAsYEyogxsEXEbMWo+SDgCEFRkxbebWU4sRyU8pRQ+Mf/9/1QhQKKAgGEAYYESog0NLYVYyBUAILwdFjX35iwzEH5OzPfcSYLHGxTfwe0KwKKAgEEAQYDyogj/U6liqyG5rWT72NZB+Rbb4kQlENp+5rrf+GXMUClsEKKAgCEAIYCyogez7o7L65H/VQnfpm9aHuQ+mZPsCTpSpdW+PjiqimxC8aLAoGL2tleS0xEiDyKsnU2cWgv94mdBHr2X9OI44H1AK4Oo3xc4XHhtytnRgB"
          }
        ]
      },
      "height": "60"
    }
  }
```
